pkg/controller/admissionchecks/provisioning/controller.go

```diff
@@ -833,3 +829,120 @@ func limitObjectName(fullName string) string {
 	hashBytes := hex.EncodeToString(h.Sum(nil))
 	return fmt.Sprintf("%s-%s", fullName[:objNameMaxPrefixLength], hashBytes[:objNameHashLength])
 }
+
+type MergedPodSet struct {
+	Name             kueue.PodSetReference
+	PodSet           *kueue.PodSet
+	PodSetAssignment *kueue.PodSetAssignment
+	Count            int32
+}
+
+func mergePodSets(
+	expectedPodSets []kueue.PodSetReference,
+	podSetMap map[kueue.PodSetReference]*kueue.PodSet,
+	psaMap map[kueue.PodSetReference]*kueue.PodSetAssignment,
+	mergeMode kueue.MergePodSetsMode,
+) []MergedPodSet {
+	merged := []MergedPodSet{}
+
+	if mergeMode == "disabled" {
+		for _, psName := range expectedPodSets {
+			ps := podSetMap[psName]
+			psa := psaMap[psName]
+			merged = append(merged, MergedPodSet{
+				Name:             psName,
+				PodSet:           ps,
+				PodSetAssignment: psa,
+				Count:            ptr.Deref(psa.Count, ps.Count),
+			})
+		}
+		return merged
+	}
+
+	for _, psName := range expectedPodSets {
+		ps := podSetMap[psName]
+		psa := psaMap[psName]
+
+		mergable := false
+		for i, mps := range merged {
+			if canMergePodSets(mps.PodSet, ps, mergeMode) {
+				merged[i].Count += ptr.Deref(psa.Count, ps.Count)
+				mergable = true
+				break
+			}
+		}
+
+		if !mergable {
+			merged = append(merged, MergedPodSet{
+				Name:             psName,
+				PodSet:           ps,
+				PodSetAssignment: psa,
+				Count:            ptr.Deref(psa.Count, ps.Count),
+			})
+		}
+	}
+
+	return merged
+}
+
+func canMergePodSets(ps1, ps2 *kueue.PodSet, mergeMode kueue.MergePodSetsMode) bool {
+	switch mergeMode {
+	case kueue.Identical:
+		// Strict comparison: all fields must match
+		return arePodSetsIdentical(ps1, ps2)
+	case kueue.Equivalent:
+		// Relaxed comparison: only resource requests must match
+		return arePodSetsEquivalent(ps1, ps2)
+	default:
+		// Unsupported merge mode
+		return false
+	}
+}
+
+func arePodSetsIdentical(ps1, ps2 *kueue.PodSet) bool {
+	if !areResourceListsEqual(ps1.Template.Spec.Resources.Requests, ps2.Template.Spec.Resources.Requests) {
+		return false
+	}
+	if !areMapsEqual(ps1.Template.Labels, ps2.Template.Labels) {
+		return false
+	}
+	if !areMapsEqual(ps1.Template.Annotations, ps2.Template.Annotations) {
+		return false
+	}
+	// What else should we compare
+	return true
+}
+
+func arePodSetsEquivalent(ps1, ps2 *kueue.PodSet) bool {
```

Comment:
For "Equivalent" I would say idendical: "ps2.Template.Spec.Resources", , "ps2.Template.Spec.Containers[].Resources", "ps2.Template.Spec.InitContainers[].Resources","ps2.Template.Spec.NodeSelector", "ps2.Template.Spec.Tolerations", "ps2.Template.Spec.Affinity".

cc @mwielgus